### ISSUE TITLE
Jupyter run error ReactorNotRestartable

### ISSUE THREAD

**Issue by liveresume** [scrapy/scrapy#2594]
I am able to run Scrapy in a Jupyter notebook.  The first time it works fine.  
However any subsequent attempts will fail with errors below.

```python
from scrapy.crawler import CrawlerProcess
 
process = CrawlerProcess({
    'USER_AGENT': 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1)'
})
process.crawl(BlogPostSpider)
process.start() # the script will block here until the crawling is finished

```
```
---------------------------------------------------------------------------
ReactorNotRestartable                     Traceback (most recent call last)
<ipython-input-8-6d30d0f44f41> in <module>()
     13 })
     14 process.crawl(spider)
---> 15 process.start() # the script will block here until the crawling is finished
     16 
     17 

/opt/conda/lib/python3.5/site-packages/scrapy/crawler.py in start(self, stop_after_crawl)
    278         tp.adjustPoolsize(maxthreads=self.settings.getint('REACTOR_THREADPOOL_MAXSIZE'))
    279         reactor.addSystemEventTrigger('before', 'shutdown', self.stop)
--> 280         reactor.run(installSignalHandlers=False)  # blocking call
    281 
    282     def _get_dns_resolver(self):

/opt/conda/lib/python3.5/site-packages/twisted/internet/base.py in run(self, installSignalHandlers)
   1240 
   1241     def run(self, installSignalHandlers=True):
-> 1242         self.startRunning(installSignalHandlers=installSignalHandlers)
   1243         self.mainLoop()
   1244 

/opt/conda/lib/python3.5/site-packages/twisted/internet/base.py in startRunning(self, installSignalHandlers)
   1220         """
   1221         self._installSignalHandlers = installSignalHandlers
-> 1222         ReactorBase.startRunning(self)
   1223 
   1224 

/opt/conda/lib/python3.5/site-packages/twisted/internet/base.py in startRunning(self)
    728             raise error.ReactorAlreadyRunning()
    729         if self._startedBefore:
--> 730             raise error.ReactorNotRestartable()
    731         self._started = True
    732         self._stopped = False

ReactorNotRestartable: 

```

**Comment by rolando** [comment]
IIRC, twisted reactor is only meant to be ran once. It cannot be restarted.

However, there is a handy package [crochet](https://github.com/itamarst/crochet) which allows to use twisted in a blocking fashion. I have played with it a while ago and published [scrapydo](https://github.com/rolando/scrapydo) but still very alpha and only covers very basic use cases. Below is an example:

```
import logging
import scrapydo
from scrapy import Request
from scrapy.linkextractors import LinkExtractor

logging.basicConfig(level=logging.DEBUG)

scrapydo.setup()

scrapydo.default_settings.update({
    'LOG_LEVEL': 'DEBUG',
    'CLOSESPIDER_PAGECOUNT': 10,
})

lx = LinkExtractor()

def parse_page(response):
    print("Got page: %s" % response.url)
    yield {'title': response.css('title::text').extract_first()}
    for link in lx.extract_links(response):
        yield Request(link.url, callback=parse_page)

scrapydo.crawl("https://www.scrapy.org", callback=parse_page)
```

**Comment by liveresume** [comment]
@rolando Thanks!  I'm experimenting with scrapy vs some custom Elixir code.  Will try yours out.

**Comment by kmike** [comment]
There is also a potential GSoC project to make "Scrapy IDE" by integrating IPython notebook event loop with Scrapy event loop; nobody started to work on it though.

Maybe we should add a note to CrawlerRunner docs with this gotcha.

**Comment by liveresume** [comment]
@kmike I don't think that scrapy should require us to jump through such a big hoop just to get it working in Jupyter.  Which gives reason to why no work has been done on it.  I realize the blame might be more on twisted, but the problem is solvable modifying only scrapy.

**Comment by kmike** [comment]
I also don't think Scrapy should require us to jump through such a big hoop just to get it working :) 

But, as you said, Scrapy is based on Twisted, and Twisted has this unfortunate limitation. Scrapy is much older than e.g. tornado or asyncio; I think using Twisted to create Scrapy was a great decision which allowed Scrapy to go on for so many years.

You can integrate Scrapy with most event-loop based apps using the CrawlerRunner API (I've done this for IPython zmq event loop and for Tornado-based HTTP API), but that's true that blocking calls are harder and require hacks/workarounds like crochet-based scrapydo. 

We do have a goal of making Scrapy more composable, and this is already improved in recent releases - CrawlerRunner API is a relatively new addition, and there is a proposed GSoC project for asyncio support (though it won't necessarily solve the problem).

**Comment by Gallaecio** [comment]
Maybe we should write a documentation page on how to use Scrapy with Jupyter Notebook for now.

**Comment by BrenBarn** [comment]
It would really be nice to have a solution to this.  I think there are a fair number of people who want to use Scrapy without thinking or knowing anything about twisted.  They just want to call `run_my_scraper()` and have it run, as many times as they want, without getting any weird errors about things like "reactors" that aren't directly relevant to use of Scrapy.  It would be great if Scrapy could incorporate any needed workarounds to provide a ready-made solution that doesn't require the user to know anything about twisted.

**Comment by rmax** [comment]
@BrenBarn it's really tricky to run twisted's reactor in a sync fashion. You can take a look at https://github.com/rmax/scrapydo (it may be outdated, though)

**Comment by nareto** [comment]
has any work been done on this? How am I supposed to develop a scrapy class in a notebook if for each modification I have to restart the kernel?

**Comment by 3reality-wangzhijie** [comment]
```
async def crawl():
    process = CrawlerProcess()
    process.crawl(BlogPostSpider)

asyncio.get_event_loop().create_task(crawl())
```

create a coroutine to run

**Comment by wRAR** [comment]
@3reality-wangzhijie this cannot solve the underlying problem.

**Comment by wRAR** [comment]
This is possible since Scrapy 2.15.0 thanks to the reactorless mode. You need to use top-level awaits instead of a synchronous `process.start()` however.
